hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java

```diff
@@ -304,32 +330,57 @@ private long acquireLock() throws UnknownHostException, TException, InterruptedE
         System.getProperty("user.name"),
         InetAddress.getLocalHost().getHostName());
     LockResponse lockResponse = metaClients.run(client -> client.lock(lockRequest));
-    LockState state = lockResponse.getState();
+    AtomicReference<LockState> state = new AtomicReference<>(lockResponse.getState());
     long lockId = lockResponse.getLockid();
 
     final long start = System.currentTimeMillis();
     long duration = 0;
     boolean timeout = false;
-    while (!timeout && state.equals(LockState.WAITING)) {
-      lockResponse = metaClients.run(client -> client.checkLock(lockId));
-      state = lockResponse.getState();
 
-      // check timeout
-      duration = System.currentTimeMillis() - start;
-      if (duration > lockAcquireTimeout) {
+    if (state.get().equals(LockState.WAITING)) {
+      try {
+        Tasks.foreach(lockId)
+            .retry(Integer.MAX_VALUE - 100) // Endless retries bound by timeouts. Tasks.retry adds 1 for "first try".
+            .exponentialBackoff(
+                lockCheckMinWaitTime,
+                lockCheckMaxWaitTime,
+                lockAcquireTimeout,
+                lockCheckBackoffScaleFactor)
+            .throwFailureWhenFinished()
+            .onlyRetryOn(WaitingForLockException.class)
+            .run(id -> {
+              try {
+                LockResponse response = metaClients.run(client -> client.checkLock(id));
+                LockState newState = response.getState();
+                state.set(newState);
+                if (newState.equals(LockState.WAITING)) {
+                  throw WAITING_FOR_LOCK_EXCEPTION;
+                }
+              } catch (InterruptedException | TException e) {
```

Comment:
For `InterruptedException`, why not throw `WaitingForLockException` and signal that the thread was interrupted? Then this could use the checked exception call, `run(id -> {...}, TException.class)` and would not need to wrap the exceptions.

Reply:
The code looks better after this comment. However throwing `WaitingForLockException` ends up losing the source of the original `InterruptedException` because it would get handled here: https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/util/Tasks.java#L452

So, I chose to throw `RuntimeException` which will stop the retry and preserve the source stack trace.

Reply:
I'd probably opt to suppress the interrupt and let the code carry on after setting that the thread was interrupted. That results in a `CommitFailedException`. I don't think that preserving the stack of the `InterruptedException` is really needed, but I'm fin with it this way if you prefer it.

Reply:
Taken care as per [the other comment](https://github.com/apache/iceberg/pull/1873#discussion_r539077360).